Spine-project/Spine-Toolbox
# Looping

This and its companion PR Spine-project/Spine-Engine#39 together implement loops in Spine Toolbox.

In Toolbox, it is now possible to connect project items with a special *Loop link* that creates a conditional jump in the dag. Currently, loop condition (continue or break?) is implemented as a user given Python script which should allow for maximum flexibility. The script is also somewhat cumbersome and counterintuitive to use so more work on this is expected in the future but not in scope of this PR.

If Engine is supplied with jumps, it will break the DAG into *chunks* at jump end points that are executed one after the other. A loop is a chunk that is executed as many times as its condition dictates (but always at least once).

**Todo:**

- [x] Implement nested loops

**Limitations**

It is not possbile to create overlapping loops except nested ones.

**Benchmarking**

A project with ~40 Data connections forming a branching DAG was used to gauge the performance impact when no loops are present. Since we basically use the same mechanism for such DAGs as we do on `master` no impact was expected. Turns out there was some and that is due to dagster 0.11 internals being noticeably slower than in 0.10.

**To test:**

Note, that you'll need to upgrade dagster to 0.11 for this!

1. Checkout this and perhaps Spine-project/Spine-Engine#39 branch
2. Upgrade requirements, should upgrade dagster as well
3. Start Toolbox
4. Check that your existing projects execute as before

1. Create Loop links between project items by holding **Alt** key while clicking the source items (loop's end point) connector button
2. Click a Loop link and edit its condition script. E.g. to execute a loop a given number of times, you can try the following:

```python
import sys
counter = int(sys.argv[1])  # Loop counter is supplied as first command line argument
exit(0 if counter < 3 else 1)  # Exit code 0 = continue, anything else = break
```

7. Execute the project. You should see from item's Execution logs that it gets executed multiple times
8. Try breaking looping by creating crazy Loop links or condition scripts

Resolves #329,
